spring-kafka/src/main/java/org/springframework/kafka/listener/KafkaMessageListenerContainer.java

```diff
@@ -1358,6 +1386,17 @@ private void commitIfNecessary() {
 			return highestOffsetMap.values();
 		}
 
+		private Collection<ConsumerRecord<K, V>> getHighestOffsetRecords(ConsumerRecords<K, V> records) {
+			final Map<TopicPartition, ConsumerRecord<K, V>> highestOffsetMap = new HashMap<>();
+			Iterator<ConsumerRecord<K, V>> iterator = records.iterator();
+			while (iterator.hasNext()) {
+				ConsumerRecord<K, V> r = iterator.next();
+				highestOffsetMap.compute(new TopicPartition(r.topic(), r.partition()),
```

Comment:
Can't really `ConsumerRecords.partitions()` help us here somehow?

Reply:
We'd still have to iterate but, yes, it would avoid having to create new TPs.

Good catch!